web/pandas/pdeps/0017-backwards-compatibility-and-deprecation-policy.md

```diff
@@ -0,0 +1,86 @@
+- Breaking changes should only occur in major releases.
+- No deprecations should be introduced in patch releases.
+
+Some bug fixes may require breaking backwards compatibility. In these cases, a deprecation cycle is not necessary. However, bug fixes which have a large impact on users might be treated as a breaking change. Whether or not a change is a bug fix or an API breaking change is a judgement call.
+
+## Deprecation process
+
+Deprecation provides a way to warn developers and give them time to adapt their code to the new functionality before the old behavior is eventually removed.
+
+A deprecation's warning message should:
+- Provide information on what is changing.
+- Mention how to achieve similar behavior if an alternative is available.
+- Include the version in which the deprecation will be enforced.
+- For large-scale deprecations, it is recommended to include a reason for the deprecation, alongside a discussion link to get user feedback.
+
+Additionally, when one introduces a deprecation, they should:
+- Use the appropriate warning class. More info on this can be found below.
+- Add the GitHub issue/PR number as a comment above the warning line.
+- Add an entry in the release notes.
+- Mention that the functionality is deprecated in the documentation using the ``.. deprecated::`` directive.
+
+### Which warning class to use
+
+Deprecations should initially use ``DeprecationWarning``, and then be switched to ``FutureWarning`` for broader visibility in the final minor version before the major release they are planned to be removed in.  
+This implementation detail can be ignored by using the appropriate ``PandasDeprecationWarning`` variable, which will be aliased to the proper warning class based on the pandas version.
+
+Not all deprecations have to use ``DeprecationWarning`` but all deprecations should eventually transition to ``FutureWarning``, i.e. deprecations in the final minor version which are planned to be removed in the major release after will immediately use ``FutureWarning``.
+
+It is recommended to not introduce large-scale deprecations in the final minor version which are planned to be removed in the major release after, since that will immediately be using a loud ``FutureWarning`` with not much time between deprecation and removal. Instead, a ``DeprecationWarning`` should be used, and the removal should be scheduled for a later major release.
+
+### Support window for the final minor version
+
+Defining a release policy is not part of this PDEP's scope, but, the final minor version plays a special role in deprecations, since:
+- It is the version where deprecations planned for removal in the next major release get switched from a ``DeprecationWarning`` to a more visible ``FutureWarning``.
+- ``FutureWarning`` deprecations released in it will immediately be removed in the next major release.
+
+And so, this document recommends a minimum of 6 months between the final minor version and the major release after it.
```

Comment:
I would take this out, the release cadence was around every 3 months for most of the 2.x cycle (even though it's been a little quiet as of right now)

Reply:
It makes sense to make sure deprecations have a minimum amount of time before being removed. I suppose we can specify the minimum duration on the deprecations themselves and not pandas versions, something like: "don't enforce any deprecation before a minimum duration of xxx".

Ping @Dr-Irv since it was his suggestion.

Reply:
@lithomas1 I suggested the 6 months between minor and major releases, because if we deprecate something in a minor release for the first time, and then do a major release 3 months later, there is not enough time for people to react to the deprecations, since a deprecation in _any_ minor release would be implemented in the next major release.

Reply:
Sometime in the past, I remember us discussing removing deprecations introduced in the last minor release in the major release after the next major release.

Did we decide to go against that?
(thinking about this again, I think I would prefer that. Deprecations deprecated in the last minor release before a major release tend to not be too important anyways)

Reply:
I'm not sure that a decision was made, or, if it was, we never wrote it down.  So the goal of this PDEP is to come to agreement on something.

Reply:
> the release cadence was around every 3 months for most of the 2.x cycle (even though it's been a little quiet as of right now)

FWIW, I think generally our release cycle has always been around every 6 months for a new minor or major release (indeed now for 2.3/3.0, it has been slower, but 2.0/2.1/2.2 has not been much faster)

But to the original point here, I do think that it is fine to say that if we introduce a deprecation in the last minor release of a series, we should not yet remove it in the next major release if that major release would happen within 6 months. 

Although I would personally also be fine with making it simpler and saying that such deprecations are just never already enforced in the next major release. I.e that is like saying that a deprecation should always be present in at least two releases. 

And that would then also make it easier to say that we always go from DeprecationWarning to FutureWarning (which would not be possible if a deprecation only exists in one release)